sbomtool/internal/deduplication/deduplicator.go

```diff
@@ -0,0 +1,304 @@
+	if len(p.CPEs) > 0 {
+		return p.CPEs[0].Attributes.String()
+	}
+
+	// Fallback to name+version+type for packages without CPE
+	name := strings.ToLower(strings.TrimSpace(p.Name))
+	version := normalizeVersion(p.Version)
+	pkgType := string(p.Type)
+
+	return fmt.Sprintf("fallback:%s|%s|%s", name, version, pkgType)
+}
+
+// normalizeVersion normalizes version strings for consistent matching.
+func normalizeVersion(version string) string {
+	if version == "" {
+		return "unknown"
+	}
+	return strings.ToLower(strings.TrimSpace(version))
+}
+
+// mergePackages merges multiple packages into a single canonical package.
+func mergePackages(packages []*pkg.Package) *pkg.Package {
+	if len(packages) == 1 {
+		return packages[0]
+	}
+
+	// Start with the first package as base
+	canonical := *packages[0]
+	canonical.SetID() // Generate new ID for merged package
+
+	// Merge CPEs (union, deduplicated)
+	cpeSet := make(map[string]cpe.CPE)
+	for _, p := range packages {
+		for _, c := range p.CPEs {
+			cpeSet[c.Attributes.String()] = c
+		}
+	}
+	canonical.CPEs = make([]cpe.CPE, 0, len(cpeSet))
+	for _, c := range cpeSet {
+		canonical.CPEs = append(canonical.CPEs, c)
+	}
+
+	// Merge licenses (union, deduplicated, skip UNKNOWN/NOASSERTION)
+	licenseSet := make(map[string]pkg.License)
+	for _, p := range packages {
+		for _, license := range p.Licenses.ToUnorderedSlice() {
+			if !isUnknownOrNoAssertion(license.Value) {
+				licenseSet[license.Value] = license
+			}
+		}
+	}
+	// If no valid licenses found, keep one UNKNOWN/NOASSERTION if it exists
+	if len(licenseSet) == 0 {
+		for _, p := range packages {
+			for _, license := range p.Licenses.ToUnorderedSlice() {
+				if isUnknownOrNoAssertion(license.Value) {
+					licenseSet[license.Value] = license
+					break
+				}
+			}
+			if len(licenseSet) > 0 {
+				break
+			}
+		}
+	}
+	licenses := make([]pkg.License, 0, len(licenseSet))
+	for _, license := range licenseSet {
+		licenses = append(licenses, license)
+	}
+	canonical.Licenses = pkg.NewLicenseSet(licenses...)
+
+	// Use non-empty version if available, otherwise keep unknown if all are unknown
+	foundValidVersion := false
+	for _, p := range packages {
+		if p.Version != "" && !isUnknownOrNoAssertion(p.Version) {
+			canonical.Version = p.Version
+			foundValidVersion = true
+			break
+		}
+	}
+	if !foundValidVersion {
+		for _, p := range packages {
+			if p.Version != "" {
+				canonical.Version = p.Version
+				break
+			}
+		}
+	}
+
+	return &canonical
+}
+
+// isUnknownOrNoAssertion checks if a value represents unknown or no assertion.
+func isUnknownOrNoAssertion(value string) bool {
+	normalized := strings.ToLower(strings.TrimSpace(value))
+	return normalized == "unknown" || normalized == "noassertion" || normalized == ""
+}
+
+// mergeOverlappingCPEGroups merges package groups that have overlapping CPEs
+func mergeOverlappingCPEGroups(keyToPackages map[string][]*pkg.Package) map[string][]*pkg.Package {
```

Comment:
similar to a comment from @arnaldo2792 , it would be nice to have a unit test for this method both for test coverage and as an example.